### DISCUSSION TITLE
C++ and WebGL

### DISCUSSION THREAD

**Discussion by FuXiii** [google/filament#6373] [Q&A] [answered]
`Filament` is mainly developed with `C++`, I had only used `Host Graphic API` like `OpenGL` and `Vulkan` and had not experience on `Web` like `WebGL` and `WebGPU`, as I know `WebGL` and `WebGPU` based in `JavaScript`, How to combine them? like use `C++` API to call `JavaScript` API?

**Comment by romainguy** [comment, marked as answer]
The WebGL version of Filament can be used with JavaScript. You could also stick with C++ and use Emscripten to compile to WASM (which is what we do).

**Comment by FuXiii** [comment]
I had researched some docs about `Emscripten`, it look like a compiler, compile `C/C++` code into `JavaScript`. Can I directly use `Emscripten` to compile pure `C/C++` `Vulkan` based API project? As far as I know, `Vulkan Standard` didn't had some `Surface` like `WebSurface`, maybe this isn't the way How to use `Emscripten`

---
If I use `Emscripten` to compile `C/C++` code into `WebGPU` with `JavaScript` code

I guess the `C/C++` code maybe like:
```CXX
#include<some webgpu implement header file>
void main()
{
  WebGPU::Navigator navigator;
  WebGPU::NavigatorGPU gpu = navigator.gpu;
  WebGPU::GPUAdapter gpu_adapter = gpu.requestAdapter();
  ...
}
```
And then use `Emscripten` to compile it into `WebGPU` code. Is that right?
And where is the `#include<some webgpu implement header file>` come from?
I had found [google:dawn](https://dawn.googlesource.com/dawn) project. Is this the way to use this project？
